NDData and NDDataBase Refactor

MSeifert04 opened the pull request:
This is the refactor of ``NDDataBase`` and ``NDData`` (from #4234). More comments will follow as soon as I get the splitted parts to work.

**edit 12/22/2015 to add checklist for completion:**

+ [x] Update documentation to reflect changes
+ [x] Remove restriction on type of `uncertainty_type` (ping list notifying about this)
+ [x] Always return the uncertainty object, not the data contained in that object, even if an `UnknownUncertainty` is created by `NDData`

MSeifert04 commented:
Summary:
``__init__``:

- New optional ``copy`` parameter. The default is ``False`` which was the case before this PR.
- Implicit parameters (i.e. ``data`` parameter is another ``NDData`` instance) are always checked. This might have produced Bugs if subclasses had ``lesser`` requirements.
- ``data`` can now be a ``Masked_Quantities`` instance
-  ``data`` with ``mask`` is only allowed if they have also a ``data`` attribute (Duck typing ``MaskedArrays``)
- If an implicit and explicit parameter is given (i.e. ``data`` was a ``Quantity`` and a ``unit`` was given) the implicit one is ignored! Before this was now always done, i.e. two different units raised an ``Exception``

``uncertainty``

- The ``uncertainty setter`` of ``NDDataBase`` has moved to ``NDData``
- ``uncertainties`` get their ``parent_nddata`` attribute set in the setter. (see issue #4152)
- The ``uncertainty setter`` now checks for an attribute ``uncertainty_type`` and if it is not present or not a string the uncertainty is wrapped in an ``UnknownUncertainty``. (Before it raised an ``Exception``)
- The getter ``uncertainty`` special cases ``UnknownUncertaintiy`` and returns only what was initially wanted as uncertainty but the internally stored ``_uncertainty`` is of ``unknown uncertainty_type``.


I think this is mostly added functionality and should be backwards compatible, except for the ``uncertainty setter``. Subclasses of ``NDDataBase`` might encounter some transitional problems since the one is now gone. All other changes only lessened restrictions (Warning instead of Exceptions).

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -192,12 +292,76 @@ def mask(self, value):
+            return self._uncertainty
+
+    @property
+    def uncertainty_type(self):
```
Is the idea here that eventually every `NDData` object would have an uncertainty, defaulting to `UnknownUncertainty`, eventually? Otherwise it seems like leaving the `uncertainty_type` a property of the uncertainty would be easier.

mwcraig replied:
Ah, right, `UnknownUncertainty` is at the top of the file.

MSeifert04 replied:
The ``uncertainty_type`` property was just a convenience I used while I tested the ``UnknownUncertainty``. I thought I deleted it before pushing but seems like I forgot.

mwcraig commented:
@wkerzendorf -- could you please take a look at this one? Looks mostly straightforward to me, the one big change is to wrap the uncertainty in an `UnknownUncertainty` if the uncertainty does not have an attribute `uncertainty_type`.

mwcraig commented:
@MSeifert04 -- once again, thanks. I gave this a quick look tonight and didn't see any problems with the code, *but* I'm not sure about dropping the requirement that the uncertainty have an `uncertainty_type` (primarily because the APE currently says it should be there).

It sounds, though, like your solution is to wrap an uncertainty that does not have an `uncertainty_type` in an `UnkownUncertainty`, right? So in the end there is always an `uncertainty_type`?

MSeifert04 commented:
@mwcraig - Yes, your summary about the ``UnknownUncertainty`` is correct. 
But note that I also changed the ``uncertainty getter`` so that for ``UnknownUncertainty`` it doesn't return the instance but only the saved ``array`` of the instance. So a normal user would not suspect that the uncertainty was modified (except in some cases, i.e. the uncertainty was a list and cast to a ``np.ndarray``). It felt a bit like cheating to be honest and if you don't like it I can revert back to the former ``enforced requirement``.

I got the idea from #3714 where @wkerzendorf proposed just such an uncertainty and I modified it a bit to match my ``NDUncertainty`` changes.

mwcraig commented:
I'll have some time tomorrow to take a longer look at the others, and to put a call out on the astropy-dev list for code reviews.

I know a bunch of people are at ADASS this week...

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -192,12 +292,76 @@ def mask(self, value):
+        saved in _uncertainty has an uncertainty_type but it's unknown :-)
+        """
+        if isinstance(self._uncertainty, UnknownUncertainty):
+            return self._uncertainty.array
```
Recommend switching this back to `return self._uncertainty`

MSeifert04 commented:
In the Hangouts meeting @wkerzendorf and @mwcraig suggested some changes. The current push reflects these:
- UnknownUncertainty is handled like any other uncertainty when accessed via the ``uncertainty`` property.
- ``uncertainty_type`` needs to be present for ``uncertainty.setter`` but we dropped the requirement for it to be a string.

I skipped the documentation update for now.

mwcraig commented:
@MSeifert04 -- could you please go ahead and do the documentation changes too (should be minimal, I think, for this one)?

MSeifert04 commented:
@mwcraig - I edited the documentation files to reflect the changes and included a section with a summary of the changes. I'm not very happy with the wording, just let me know if anything should be edited.

mwcraig commented on docs/nddata/index.rst:
```diff
@@ -156,6 +157,57 @@ Code that uses the arithemtic methods that used to be included in
+
+`~astropy.nddata.NDDataBase`:
+
++ `NDDataBase.uncertainty.getter`: is now abstract.
```
This and the line below are causing the sphinx failures because there is no `getter` or `setter` method.

mwcraig replied:
Won't be an issue, though, if this just moves to the changelog.

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -5,170 +5,284 @@
+        uncertainty_type: ``"unknown"``
+            The type of uncertainty is unknown.
+
+        Notes
```
Could remove this note (or move it to a comment in the class body).

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -5,170 +5,284 @@
+        # Check if meta is a dict and create an empty one if no meta was given
+        if meta is None:
+            meta = OrderedDict()
+        elif not isinstance(meta, collections.Mapping):
```
nifty - hadn't realized `collections.Mapping` could be used to check for a dict-like interface.

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -5,170 +5,284 @@
-        The actual data contained in this `NDData` object. If possible, data
-        will not be copied`data`, so you should make copy the ``data`` before
-        passing it in if that's the desired behavior.
+    data : `~numpy.ndarray`-like, `NDData`-like or list
```
Can drop list, it is array-like already.

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -5,170 +5,284 @@
+                 meta=None, unit=None, copy=False):
 
+        # Rather pointless since the NDDataBase does not implement any setting
+        # but before (#4234) the NDDataBase did call the uncertainty
```
Drop the reference to #4234

mwcraig commented on astropy/nddata/nddata_base.py:
```diff
@@ -14,79 +14,66 @@
-    classes.
+    NDDataBase makes no attempt of restricting any attribute or defines how
+    these are stored. Therefore all properties and methods have to be
+    overridden in subclasses. See `~astropy.nddata.NDData` for a subclass that
```
Suggested edit: drop the sentence that begins "NDDataBase makes...". 

Change "Therefore all" -> "All"

mwcraig commented on astropy/nddata/tests/test_nddata.py:
```diff
@@ -11,31 +11,41 @@
-from ...tests.helper import pytest, raises
+from ...utils.compat.odict import OrderedDict
+from ..nduncertainty import StdDevUncertainty  # , NDUncertainty
+from ...tests.helper import pytest  # , raises
```
Please drop the unused imports completely rather than making them comments (and thanks for the clean-up).

mwcraig commented on astropy/nddata/tests/test_nddata.py:
```diff
@@ -57,27 +67,226 @@ def __getitem__(self):
+        assert ndd.data[1] != quantity.value[1]
+
+
+def test_nddata_init_data_maskedQuantity():
```
Pedantic naming issue: `maskedQuatity` -> `masked_quantity`

mwcraig commented on astropy/nddata/tests/test_nddata.py:
```diff
@@ -57,27 +67,226 @@ def __getitem__(self):
+
+    # Now let's see what happens if we have all explicitly set
+    nd1 = NDData(np.array([1]), mask=False, uncertainty=10, unit=u.s,
+                 meta={'dest': 'mordor'}, wcs=10)
```
Nice meta :)

mwcraig commented on astropy/nddata/tests/test_nddata.py:
```diff
@@ -57,27 +67,226 @@ def __getitem__(self):
+    assert nd3.meta != nd1.meta
+
+
+def test_nddata_init_data_nddataSubclass():
```
`nddataSubclass` -> `nddata_subclass`

mwcraig commented on astropy/nddata/tests/test_nddata_base.py:
```diff
@@ -5,7 +5,7 @@
 
 from ..nddata_base import NDDataBase
-from ...tests.helper import pytest
+# from ...tests.helper import pytest
```
Remove instead of commenting out if unneeded.

mwcraig commented on docs/nddata/index.rst:
```diff
@@ -24,7 +24,8 @@ datasets in astropy through:
   `~astropy.nddata` has changed significantly in astropy 1.0. See the section
-  :ref:`nddata_transition` for more information.
+  :ref:`nddata_transition` for more information. Some more changes were done in
+  astropy x.xx, see more details in :ref:`nddata_transition2`.
```
I think we could take out this warning box altogether...it has been (almost?) a year since 1.0 was released.

mwcraig commented on docs/nddata/index.rst:
```diff
@@ -156,6 +157,57 @@ Code that uses the arithemtic methods that used to be included in
+.. _nddata_transition2:
+
+Transition to astropy x.xx
+==========================
```
I think this should be in the changelog rather than a separate section in the sphinx documentation. All of the detailed changes show up in the docstrings and examples, and

mwcraig commented:
@MSeifert04 -- looks really good, just a few minor inline comments. Most of the bullet points you added to `index.rst` should go in `CHANGES.rst` instead -- please put them in the section for version 1.2.

mwcraig commented on docs/nddata/index.rst:
```diff
@@ -156,6 +157,57 @@ Code that uses the arithemtic methods that used to be included in
+    by default False. If it is True all other parameters are copied before saving
+    them as attributes. If False the parameters are only copied if there is no
+    way of saving them as reference.
+  + the ``data`` parameter allows now also ``Masked_Quantity`` objects
```
I don't think `Quantity` can be masked yet?

mwcraig commented on docs/nddata/index.rst:
```diff
@@ -156,6 +157,57 @@ Code that uses the arithemtic methods that used to be included in
+  + if the ``data`` is another instance or subclass of
+    `~astropy.nddata.NDData` all implicit properties are checked because
+    subclasses might implement another set of restrictions.
+  + if an implicit and explicit attribute is provided, the
```
This point should be added either as an example in getting started (show an example of what happens if you, say, already have a unit but supply a new one when creating a new NDData) or in `nddata.rst`

mwcraig commented on docs/nddata/index.rst:
```diff
@@ -156,6 +157,57 @@ Code that uses the arithemtic methods that used to be included in
+    explicit one is kept (without conversion) and a warning is issued. For
+    example if ``data`` is a `~astropy.units.Quantity` and also a ``unit`` is
+    given. The unit of the instance will be the ``unit`` parameter.
+  + allowed a ``data`` parameter with a ``mask`` without checking that it had a
```
This I think doesn't need to be mentioned in the changelog unless there is an open issue for it.

MSeifert04 commented:
@mwcraig - Thanks for all the input. 

I changed it accordingly but some comments are now outdated but I'm not sure how to deal with them:

- I dropped not only the warning box in the ``docs/nddata/index.rst`` but also the section about the transition. I'm not sure if you meant that when you said we needed to remove the warning.
- You mentioned ``MaskedQuantity``: As far as I know it is not stabilized but there are ways of creating such objects by masking a quantity. On the other hand multiplying a masked array with a unit doesn't create one.

- I'm not sure if the changes in ``Changes.rst`` are in the right subsections, it's quite hard to decide if something is a bugfix, an api change or a new feature but I tried my best.

mwcraig commented on CHANGES.rst:
```diff
@@ -32,6 +32,14 @@ New Features
 
+  - ``NDData.__init__``: got an additional optional parameter ``copy``
+    which is False by default. If it is True all other parameters are copied
+    before saving them as attributes. If False the parameters are only copied
```
A few minor changes:

+ ``NDData.__init__`` -> ``NDData`` 
+ "got an additional" -> "added an"
+ Add the issue number in brackets at the end: `[#4270]`

mwcraig commented on CHANGES.rst:
```diff
@@ -32,6 +32,14 @@ New Features
+    before saving them as attributes. If False the parameters are only copied
+    if there is no way of saving them as reference.
+
+  - ``NDData.__init__``: ``Masked_Quantity`` objects are allowed as ``data``
```
+ NDData.__init__ -> NDData
+ Add the issue number in brackets at the end: `[#4270]`

mwcraig commented on CHANGES.rst:
```diff
@@ -81,6 +89,26 @@ API changes
+    - did not set the ``parent_nddata`` of the ``uncertainty`` if the uncertainty
+      was a subclass of ``.NDUncertainty``. #4152
+
+  - ``NDData.__init__``: if the ``data`` is another instance or subclass of
```
This one could be dropped I think. Doesn't really change the intended behavior, and it had not been reported as a bug.

mwcraig commented on CHANGES.rst:
```diff
@@ -81,6 +89,26 @@ API changes
 
 - ``astropy.nddata``
 
+  - ``NDDataBase`` only defines an abstract ``uncertainty`` getter instead of
```
How about combining this and the next one into one, phrased something like: ``NDDataBase`` does not implement a setter or getter for ``uncertainty``, which is now an `abstractproperty`.

And add the PR # in brackets.

mwcraig commented on CHANGES.rst:
```diff
@@ -81,6 +89,26 @@ API changes
+  - ``NDDataBase`` doesn't implement an ``uncertainty`` setter. It was moved to
+    ``NDData``. The setter was also modified:
+
+    - it wraps an ``uncertainty`` inside an ``UnknownUncertainty``
```
Make each of these sub-bullets main level bullets. So:

`NDData` wraps the `uncertainty` inside an `UnknownUncertainty`...

and add the issue number.

mwcraig commented on CHANGES.rst:
```diff
@@ -81,6 +89,26 @@ API changes
+      if no ``uncertainty_type`` attribute is present in the uncertainty instead
+      of raising an Exception.
+
+    - the uncertainty_type of the uncertainty must not be a string
```
+ Capitalize "the"
+ Double backticks around `uncertainty_type`
+ "must not be anymore." -> "is no longer required to be, but..."

mwcraig replied:
@eteq -- this is a change from what was originally described in the APE but seems sensible to @wkerzendorf and I. IIRC you had strong feelings about this so wanted to point it out.

mwcraig commented on CHANGES.rst:
```diff
@@ -81,6 +89,26 @@ API changes
+    - the uncertainty_type of the uncertainty must not be a string
+      anymore. But it is still recommended.
+
+    - did not set the ``parent_nddata`` of the ``uncertainty`` if the uncertainty
```
Unclear...is it that the `parent_nddata` was not properly set before but now it is? If so, maybe: "The `parent_ndddata` of the `uncertainty` is now set regardless of the class of the `uncertainty`"?

Also, the issue number should be in brackets.

mwcraig commented:
@MSeifert04 -- looking good, just some minor changes in the changelog entries.

MSeifert04 commented:
Apart from the changes, I just found that there is a descriptor lingering in the ``utils`` for the ``meta``: https://github.com/astropy/astropy/blob/master/astropy/utils/metadata.py#L115

Since this is completly identical to the meta handling (except that that it implements a setter which isn't present until now) in ``NDData``: Should I include it here?

Would look something like this
```
from astropy.utils.metadata import MetaData
class NDData(NDDataBase):
    meta = MetaData()
    ...
```

wkerzendorf commented:
I think this is a very good idea. and you can subclass this to make the arithmetic operations on them.

wkerzendorf commented:
But this PR does not deal with meta, right? Did we merge one of them?

mwcraig commented:
@Mseifert04 -- could you please open a separate issue to remind us to look at moving metadata to the one in utils? 

That way we can get this merged sooner but still remember we need to come back to the issue.

MSeifert04 commented:
@mwcraig I've opened a seperate issue and will make a seperate PR as soon as this is merged.

mwcraig commented:
@MSeifert04 -- looks like this just needs a rebase (probably related to the changelog) then it is ready to merge.

MSeifert04 commented:
@mwcraig - My first rebase - I hope I did not screw it up. The rebase was because of #4466.

mwcraig commented on astropy/nddata/tests/test_nddata.py:
```diff
@@ -94,6 +303,101 @@ def __repr__(self):
+    nd = NDData([1, 2, 3], meta={})
+    assert len(nd.meta) == 0
+    nd = NDData([1, 2, 3])
+    assert isinstance(nd.meta, OrderedDict)
```
This is where the test failures are happening:

```python
_______________________________ test_param_meta ________________________________
    def test_param_meta():
        # everything dict-like is allowed
        with pytest.raises(TypeError):
            NDData([1], meta=3)
        nd = NDData([1, 2, 3], meta={})
        assert len(nd.meta) == 0
        nd = NDData([1, 2, 3])
>       assert isinstance(nd.meta, OrderedDict)
E       assert isinstance(OrderedDict(), OrderedDict)
E        +  where OrderedDict() = NDData([1, 2, 3]).meta
```

mwcraig commented on astropy/nddata/tests/test_nddata.py:
```diff
@@ -11,31 +11,41 @@
 from ..nddata import NDData
-from ..nduncertainty import StdDevUncertainty, NDUncertainty
-from ...tests.helper import pytest, raises
+from ...utils.compat.odict import OrderedDict
```
Am guessing this needs to change now that the compat odict is gone?

astrofrog replied:
Yes

MSeifert04 replied:
Alright, I'll fix it later. Thanks for tracking it down.

MSeifert04 replied:
@mwcraig and @astrofrog Thanks - I've pushed a fix.

mwcraig commented:
Note to self: also checked this with ccdproc and nothing broke...

mwcraig commented:
:tada: Thanks again for this @MSeifert04!
